azure/services/privatelinks/spec.go

```diff
@@ -0,0 +1,239 @@
+		if ipAllocationMethod != network.Dynamic && ipAllocationMethod != network.Static {
+			return nil, errors.Errorf("%T is not a supported network.IPAllocationMethodStatic", natIPConfiguration.AllocationMethod)
+		}
+		var privateIpAddress *string
+		if ipAllocationMethod == network.Static {
+			if natIPConfiguration.PrivateIPAddress != "" {
+				privateIpAddress = pointer.String(natIPConfiguration.PrivateIPAddress)
+			} else {
+				return nil, errors.Errorf("Private link NAT IP configuration with static IP allocation must specify a private address")
+			}
+		}
+		ipConfiguration := network.PrivateLinkServiceIPConfiguration{
+			Name: pointer.String(fmt.Sprintf("%s-natipconfig-%d", natIPConfiguration.Subnet, i+1)),
+			PrivateLinkServiceIPConfigurationProperties: &network.PrivateLinkServiceIPConfigurationProperties{
+				Subnet: &network.Subnet{
+					ID: pointer.String(
+						fmt.Sprintf(
+							"/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/virtualNetworks/%s/subnets/%s",
+							s.SubscriptionID,
+							s.ResourceGroup,
+							s.VNet,
+							natIPConfiguration.Subnet)),
+				},
+				PrivateIPAllocationMethod: ipAllocationMethod,
+				PrivateIPAddress:          privateIpAddress,
+			},
+		}
+		ipConfigurations = append(ipConfigurations, ipConfiguration)
+		ipConfigurations[0].Primary = pointer.Bool(true)
+	}
+
+	// Load balancer front-end IP configurations
+	var frontendIPConfigurations []network.FrontendIPConfiguration
+	for _, frontendIPConfigName := range s.LBFrontendIPConfigNames {
+		frontendIPConfig := network.FrontendIPConfiguration{
+			ID: pointer.String(
+				fmt.Sprintf("/subscriptions/%s/resourceGroups/%s/providers/Microsoft.Network/loadBalancers/%s/frontendIPConfigurations/%s",
+					s.SubscriptionID,
+					s.ResourceGroupName(),
+					s.LoadBalancerName,
+					frontendIPConfigName)),
+		}
+		frontendIPConfigurations = append(frontendIPConfigurations, frontendIPConfig)
+	}
+
+	privateLinkToCreate := network.PrivateLinkService{
+		Name:     pointer.String(s.Name),
+		Location: pointer.String(s.Location),
+		PrivateLinkServiceProperties: &network.PrivateLinkServiceProperties{
+			IPConfigurations:                     &ipConfigurations,
+			LoadBalancerFrontendIPConfigurations: &frontendIPConfigurations,
+			EnableProxyProtocol:                  s.EnableProxyProtocol,
+		},
+		Tags: converters.TagsToMap(infrav1.Build(infrav1.BuildParams{
+			ClusterName: s.ClusterName,
+			Lifecycle:   infrav1.ResourceLifecycleOwned,
+			Name:        pointer.String(s.Name),
+			Additional:  s.AdditionalTags,
+		})),
+	}
+
+	if len(s.AllowedSubscriptions) > 0 {
+		privateLinkToCreate.Visibility = &network.PrivateLinkServicePropertiesVisibility{
+			Subscriptions: &s.AllowedSubscriptions,
+		}
+	}
+	if len(s.AutoApprovedSubscriptions) > 0 {
+		privateLinkToCreate.AutoApproval = &network.PrivateLinkServicePropertiesAutoApproval{
+			Subscriptions: &s.AutoApprovedSubscriptions,
+		}
+	}
+
+	if existing != nil {
```

Comment:
I see why you have this towards the end, but can we move the check to the beginning of the function? That way, it's consistent with the others, and if we get the wrong type we can fail early. In order to check if it's up to date, consider something like this:
```go
if existing != nil {
	existingPrivateLink, ok := existing.(network.PrivateLinkService)
	if !ok {
		return nil, errors.Errorf("%T is not a network.PrivateLinkService", existing)
	}

	privateLinkToCreate, err := s.Parameters(ctx, nil)
	if err != nil {
		return nil, err
	}
	
	if isExistingUpToDate(existingPrivateLink, privateLinkToCreate) {
		return nil, nil
	}
}
```

Reply:
Good point, will align it.

Reply:
Update this part of code so there is am early check of `existing` type and now it's similar to your code snipped above (with the remaining required parts of the code there).